### PR TITLE
User config-stored sqlite `obj.describe()` cache

### PR DESCRIPTION
**Pull request by klauer**
## Description
* Adds `utils.get_config_path` for getting a valid configuration path for the 3 main platforms
    - This can be overridden by the environment variable `TYPHOS_CONFIG_PATH`
* Adds `cache._DescribeDatabase`, an sqlite3-backed database for storing ophyd object descriptions.
    - Path defaults to `$TYPHOS_CONFIG_PATH/typhos_describe.sqlite`
    - This can be overridden by env variable `$TYPHOS_DATABASE_PATH`
    - `TYPHOS_DATABASE_PATH` may be `:memory:`, should issues arise
* `cache._DescribeDatabase` and `cache._GlobalDescribeCache` are loosely tied together
   - A single Qt signal connects the in-memory cache to the persistent cache
   - A single attribute in the in-memory cache queries the persistent cache

## Motivation and Context
With a persistent cache, the backing control system does not need to be running for the correct widgets to be created with typhos.

First pass at (but does not close) #285 

## How Has This Been Tested?
Locally and with a couple unit tests, provided.

Note: *This _must_ be tested for viability on NFS prior to potential merging.*

## Where Has This Been Documented?
Docstrings, so far

## TODO
- [ ] Do not cache fake signals
- [x] More explicit option to disable this caching, or is `:memory:` above sufficient?

### CODE REVIEW

**Comment by klauer** [comment]
To reiterate from the PR description, this needs testing with real devices and real machines (e.g., not personal laptops where the sqlite file is on a fast SSD).

**Comment by ZLLentz** [comment]
I'll give this a whirl on NFS land

File: typhos/utils.py [hunk 1]
```diff
@@ -1058,3 +1069,30 @@ def get_text(item):
...
+    if platform.system().lower() == 'windows':
+        path = os.getenv('APPDATA') or '~'
+    elif platform.system().lower() == 'darwin':
+        path = pathlib.Path('~') / 'Library' / 'Caches'
+    else:
+        path = os.getenv('XDG_CACHE_HOME') or '~/.cache'
+
+    # Allow TYPHOS_CONFIG_PATH to override the above operating system-specific
+    # entries:
+    path = os.environ.get('TYPHOS_CONFIG_PATH', path / 'typhos')
```
Comment (ZLLentz) [inline typhos/utils.py:1085]: This fails on linux because path is a string, unlike macos which is pre-converted to a `pathlib.Path`

**Comment by ZLLentz** [comment]
On psbuild-rhel7-01 using the default cache location and the latest tags of everything else:

It took about 1 minute from first log message to last while writing the cache, though the screen wasn't responsive to resizing/scrolling/closing after this for a while. It also wouldn't close very quickly... I suspect something was still initializing. This is a slowdown from the previous 40 seconds, but this is expected because we're checking and writing to a cache.

On second load (reading the cache) it took about 2 minutes 40 seconds but was very responsive immediately. This suggests writing a cache is counterproductive on NFS since we used to be able to do this in 1 minute.

Third load was the same as the second load.

A quick test in IPython after the cache had been written out using
 `typhos "pcdsdevices.device_types.PPM[{'prefix':'IM3L0:PPM', 'name':'im3l0'}]"`:

```
from typhos.cache import _DescribeDatabase
from pcdsdevices.device_types import PPM

db = _DescribeDatabase()
im3l0 = PPM('IM3L0:PPM', name='im3l0')
sig = im3l0.led

In [1]: %timeit db.get(sig)
2.11 ms ± 49.6 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
In [2]: len(list(im3l0.walk_signals()))
Out[2]: 287
In [3]: def all_gets():
    ...:     for walk in im3l0.walk_signals():
    ...:         db.get(walk.item)
    ...:
In [4]: %timeit all_gets()
620 ms ± 16.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

This suggests that the database action is fine and should only amount to a little over half a second of time spent across the full device... meaning that NFS seems ok and there is some other problem?

I have not had a chance to dive into the code yet, except the minimum required to set up the `IPython` test.

File: typhos/utils.py (hunk 1 above)
Comment (klauer) [reply typhos/utils.py:1085]: D'oh - I had rewritten this like 3 times, thinking I had gotten the most simple version. Will fix.

**Comment by klauer** [comment]
Thanks for the test effort, @ZLLentz. As-is, the database-backed cache doesn't make any effort to batch reads/writes, which could significantly improve performance over NFS. 

That is: every `describe` is immediately committed (saved) into the database, and every `get` call cache miss results in a database read.

Your near 3 minute result is very confusing. I think I'll grab your database and attempt to reproduce it...

**Comment by klauer** [comment]
Force-pushed after rebasing on conflicting cache changes.

File: typhos/cache.py [hunk 2]
```diff
@@ -3,7 +3,6 @@
 import fnmatch
 import functools
 import logging
-import ophyd
```
Comment (hhslepicka) [inline typhos/cache.py]: it was me! :D my bad!

Comment (klauer) [reply typhos/cache.py]: Even worse, I pushed a commit that shouldn't be here:  52e2f7c ; re-force pushing to save face

**Comment by klauer** [comment]
Added, then removed, queueing signal work by @hhslepicka which looks promising but will require fixing up a bunch of tests; moving to a separate PR.

**Comment by klauer** [comment]
(re-requesting reviews from @hhslepicka and @ZLLentz ) 
I think this is probably good enough for now. Most of the code here is reasonably well isolated - as in it does not touch other parts of the codebase.

If there's even a bit of concern, we could perhaps make the default setting to be 'no persistent database unless specified by the environment'.

**Comment by hhslepicka** [comment]
The cache code here is solid and I think that it is a great addition to the codebase.
I would vote to merge it in so we can address the remaining parts at separated PRs.

**Comment by ZLLentz** [comment]
I'll clear my cache and give this a try again

**Comment by ZLLentz** [comment]
This still is very slow on the NFS with the PPM, slower than without the patch (2x slowdown, from 1 minute to 2 minutes). Loading from an empty cache and from a full cache both took around 2 minutes.

File: typhos/cache.py [hunk 3]
```diff
@@ -341,3 +386,291 @@ def add_path(self, path):
...
+        # Support __getitem__ in row access:
+        self._con.row_factory = sqlite3.Row
+
+        self.log.info('Describe database path: %s', path)
+        self._con.execute(self._create_query)
+
+        t0 = time.time()
+        self._preload()
+        elapsed = time.time() - t0
+        self.log.info('Took %d ms to preload the cache', int(elapsed * 1000))
```
Comment (ZLLentz) [inline typhos/cache.py:519]:
```suggestion
        self.log.debug('Took %d ms to preload the cache', int(elapsed * 1000))
```
This isn't useful information for the end-user so it belongs in debug:
```
[2020-05-07 14:42:13] - INFO - Loading "pcdsdevices.device_types.PPM[{'prefix':'IM3L0:PPM'}]" ...
[2020-05-07 14:42:15] - INFO - Loading Tools ...
[2020-05-07 14:42:15] - INFO - Describe database path: /reg/neh/home4/zlentz/.cache/typhos/typhos_describe.sqlite
[2020-05-07 14:42:15] - INFO - Took 2 ms to preload the cache
[2020-05-07 14:42:16] - INFO - Adding devices ...
[2020-05-07 14:42:42] - INFO - Launching application ...
[2020-05-07 14:44:02] - INFO - Execution complete!
```
```
[2020-05-07 14:44:20] - INFO - Loading "pcdsdevices.device_types.PPM[{'prefix':'IM3L0:PPM'}]" ...
[2020-05-07 14:44:21] - INFO - Loading Tools ...
[2020-05-07 14:44:22] - INFO - Describe database path: /reg/neh/home4/zlentz/.cache/typhos/typhos_describe.sqlite
[2020-05-07 14:44:28] - INFO - Took 6102 ms to preload the cache
[2020-05-07 14:44:28] - INFO - Adding devices ...
[2020-05-07 14:46:01] - INFO - Launching application ...
[2020-05-07 14:46:17] - INFO - Execution complete!
```

Comment (klauer) [reply typhos/cache.py]: Agreed and accepted.

However, that 6 seconds is atrocious. Perhaps that indicates this just shouldn't be merged at all (or at minimum an alternative pre-loading method should be done)

**Comment by klauer** [comment]
Yeah, I think this isn't worth merging given the NFS issues. I don't think we'll be able to get a file-based solution on our production machines due to this.

Any alternative I can think of comes with its own set of issues:
* Use a copy of the database and work on it in-memory
   - Fast startup; but commits slow + synchronization issues
* Use a non-NFS location (then it's not available everywhere to the user)

I'm going to take the tests out of this PR and close it, as at least they are useful.

**Pull request closed without merging**